http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala

```diff
@@ -1,74 +1,220 @@
-            None
-          }
-        } flatMap { messageOpt =>
-          Future.failed(CallToSamFailedException(httpRequest.uri, response.status, messageOpt))
+class HttpSamDAO(httpClient: Client[IO], config: HttpSamDao2Config)(implicit logger: Logger[IO]) extends SamDAO with Http4sClientDsl[IO] {
+  private[leonardo] val petTokenCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(config.petCacheExpiryTime.toMinutes, TimeUnit.MINUTES)
+    .maximumSize(config.petCacheMaxSize)
+    .build(
+      new CacheLoader[UserEmailAndProject, IO[Option[String]]] {
+        def load(userEmailAndProject: UserEmailAndProject): IO[Option[String]] = {
+          getPetAccessToken(userEmailAndProject.userEmail, userEmailAndProject.googleProject)
         }
       }
+    )
+
+  def getStatus(): IO[StatusCheckResponse] = {
+    httpClient.expectOr[StatusCheckResponse](Request[IO](
+      method = Method.GET,
+      uri = config.samUri.withPath(s"/status")
+    ))(onError)
+  }
+
+
+  def hasResourcePermission(resourceId: ValueObject, action: String, resourceTypeName: ResourceTypeName, authHeader: Authorization): IO[Boolean] = for {
+    res <- httpClient.expectOr[Boolean](Request[IO](
+      method = Method.GET,
+      uri = config.samUri.withPath(s"/api/resources/v1/${resourceTypeName.toString}/${resourceId.value}/action/${action}"),
+      headers = Headers.of(authHeader)
+    ))(onError)
+  } yield res
+
+  def getResourcePolicies[A](authHeader: Authorization, resourseTypeName: ResourceTypeName)(implicit decoder: EntityDecoder[IO, List[A]]): IO[List[A]] = {
+    httpClient.expectOr[List[A]](Request[IO](
+      method = Method.GET,
+      uri = config.samUri.withPath(s"/api/resources/v1/${resourseTypeName.toString}"),
+      headers = Headers.of(authHeader)
+    ))(onError)
+  }
+
+  //Notifications that Leo has created/destroyed clusters. Allows the auth provider to register things.
+  /**
+    * Leo calls this method to notify the auth provider that a new notebook cluster has been created.
+    * The returned future should complete once the provider has finished doing any associated work.
+    * Returning a failed Future will prevent the cluster from being created, and will call notifyClusterDeleted for the same cluster.
+    * Leo will wait, so be timely!
+    *
+    * @param internalId     The internal ID for the cluster (i.e. used for Sam resources)
+    * @param creatorEmail     The email address of the user in question
+    * @param googleProject The Google project the cluster was created in
+    * @param clusterName   The user-provided name of the Dataproc cluster
+    * @return A Future that will complete when the auth provider has finished doing its business.
+    */
+  def createCluster(internalId: ClusterInternalId, creatorEmail: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, traceId: TraceId = TraceId(randomUUID)): IO[Unit] = {
```

Comment:
The method name throws me off a bit since we're not actually creating a cluster with it but creating the cluster's associated auth resources within the auth provider...so I wonder if something like `createClusterAuthResources` may be clearer. Similarly for `deleteCluster()` below.

Reply:
updated